File: SignalServiceKit/src/Network/API/RESTNetworkManager.m
//
//  Copyright (c) 2022 Open Whisper Systems. All rights reserved.
//

#import "RESTNetworkManager.h"
#import "OWSQueues.h"
#import "TSRequest.h"
#import <SignalServiceKit/SignalServiceKit-Swift.h>

NS_ASSUME_NONNULL_BEGIN

// Session managers are stateful (e.g. the headers in the requestSerializer).
// Concurrent requests can interfere with each other. Therefore we use a pool
// do not re-use a session manager until its request succeeds or fails.
@interface OWSSessionManagerPool : NSObject

@property (nonatomic) NSMutableArray<RESTSessionManager *> *pool;
@property (atomic, nullable) NSDate *lastDiscardDate;

@end

#pragma mark -

@implementation OWSSessionManagerPool

- (instancetype)init
{
    self = [super init];
    if (!self) {
        return self;
    }
    
    self.pool = [NSMutableArray new];

    [NSNotificationCenter.defaultCenter
        addObserver:self
           selector:@selector(isCensorshipCircumventionActiveDidChange)
               name:OWSSignalService.isCensorshipCircumventionActiveDidChangeNotificationName
             object:nil];

    [NSNotificationCenter.defaultCenter addObserver:self
                                           selector:@selector(isSignalProxyReadyDidChange)
                                               name:SignalProxy.isSignalProxyReadyDidChangeNotificationName
                                             object:nil];

    return self;
}

- (void)isCensorshipCircumventionActiveDidChange
{
    OWSAssertIsOnMainThread();

    self.lastDiscardDate = [NSDate new];
}

- (void)isSignalProxyReadyDidChange
{
    OWSAssertIsOnMainThread();

    self.lastDiscardDate = [NSDate new];
}

- (RESTSessionManager *)get
{
    AssertOnDispatchQueue(NetworkManagerQueue());

    // Iterate over the pool, discarding expired session managers
    // until we find a unexpired session manager in the pool or
    // drain the pool and create a new session manager.
    while (YES) {
        RESTSessionManager *_Nullable sessionManager = [self.pool lastObject];
        if (sessionManager == nil) {
            // Pool is drained; make a new session manager.
            return [RESTSessionManager new];
        }
        OWSAssertDebug(sessionManager != nil);
        [self.pool removeLastObject];
        if ([self shouldDiscardSessionManager:sessionManager]) {
            // Discard.
            continue;
        }
        return sessionManager;
    }
}

- (void)returnToPool:(RESTSessionManager *)sessionManager
{
    AssertOnDispatchQueue(NetworkManagerQueue());
    
    OWSAssertDebug(sessionManager);
    const NSUInteger kMaxPoolSize = CurrentAppContext().isNSE ? 5 : 32;
    if (self.pool.count >= kMaxPoolSize || [self shouldDiscardSessionManager:sessionManager]) {
        // Discard.
        return;
    }
    [self.pool addObject:sessionManager];
}

- (BOOL)shouldDiscardSessionManager:(RESTSessionManager *)sessionManager
{
    NSDate *_Nullable lastDiscardDate = self.lastDiscardDate;
    if (lastDiscardDate != nil &&
        [lastDiscardDate isAfterDate:sessionManager.createdDate]) {
        return YES;
    }
    return fabs(sessionManager.createdDate.timeIntervalSinceNow) > self.maxSessionManagerAge;
}

- (NSTimeInterval)maxSessionManagerAge
{
    // Throw away session managers every 5 minutes.
    return 5 * kMinuteInterval;
}

@end

#pragma mark -

@interface RESTNetworkManager ()

// These properties should only be accessed on serialQueue.
@property (atomic, readonly) OWSSessionManagerPool *udSessionManagerPool;
@property (atomic, readonly) OWSSessionManagerPool *nonUdSessionManagerPool;

@end

#pragma mark -

@implementation RESTNetworkManager

- (instancetype)init
{
    self = [super init];
    if (!self) {
        return self;
    }
    
    _udSessionManagerPool = [OWSSessionManagerPool new];
    _nonUdSessionManagerPool = [OWSSessionManagerPool new];
    
    OWSSingletonAssert();
    
    return self;
}

#pragma mark Manager Methods

- (void)makeRequest:(TSRequest *)request
    completionQueue:(dispatch_queue_t)completionQueue
            success:(RESTNetworkManagerSuccess)success
            failure:(RESTNetworkManagerFailure)failure
{
    OWSAssertDebug(request);
    OWSAssertDebug(success);
    OWSAssertDebug(failure);
    
    dispatch_async(NetworkManagerQueue(), ^{
        [self makeRequestSync:request completionQueue:completionQueue success:success failure:failure];
    });
}

- (void)makeRequestSync:(TSRequest *)request
        completionQueue:(dispatch_queue_t)completionQueue
                success:(RESTNetworkManagerSuccess)successParam
                failure:(RESTNetworkManagerFailure)failureParam
{
    OWSAssertDebug(request);
    OWSAssertDebug(successParam);
    OWSAssertDebug(failureParam);
    
    BOOL isUDRequest = request.isUDRequest;
    NSString *label = (isUDRequest ? @"UD request" : @"Non-UD request");
    if (isUDRequest) {
        OWSAssert(!request.shouldHaveAuthorizationHeaders);
    }
    OWSLogInfo(@"Making %@: %@", label, request);
    
    OWSSessionManagerPool *sessionManagerPool
    = (isUDRequest ? self.udSessionManagerPool : self.nonUdSessionManagerPool);
    RESTSessionManager *sessionManager = [sessionManagerPool get];
    
    RESTNetworkManagerSuccess success = ^(id<HTTPResponse> response) {
#if TESTABLE_BUILD
        if (SSKDebugFlags.logCurlOnSuccess) {
            [HTTPUtils logCurlForURLRequest:request];
        }
#endif
        
        dispatch_async(NetworkManagerQueue(), ^{
            [sessionManagerPool returnToPool:sessionManager];
        });

        dispatch_async(completionQueue, ^{
            OWSLogInfo(@"%@ succeeded (%ld) : %@", label, response.responseStatusCode, request);

            if (request.canUseAuth && request.shouldHaveAuthorizationHeaders) {
                [RESTNetworkManager.tsAccountManager setIsDeregistered:NO];
            }

            successParam(response);
            
            [OutageDetection.shared reportConnectionSuccess];
        });
    };
    RESTNetworkManagerFailure failure = ^(OWSHTTPErrorWrapper *errorWrapper) {
        OWSAssertDebug(errorWrapper);

        dispatch_async(NetworkManagerQueue(), ^{
            [sessionManagerPool returnToPool:sessionManager];
        });

        dispatch_async(completionQueue, ^{ failureParam(errorWrapper); });
    };

    [sessionManager performRequest:request
                           success:success
                           failure:failure];
}

@end

NS_ASSUME_NONNULL_END

File: SignalServiceKit/Network/API/RESTNetworkManager.swift
//
// Copyright 2021 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation

class RESTNetworkManager {
    private let udSessionManagerPool = OWSSessionManagerPool()
    private let nonUdSessionManagerPool = OWSSessionManagerPool()

    init() {
        SwiftSingletons.register(self)
    }

    private func makeRequest(_ request: TSRequest) async throws -> any HTTPResponse {
        let isUdRequest = request.isUDRequest
        if (isUdRequest) {
            owsPrecondition(!request.shouldHaveAuthorizationHeaders)
        }

        let sessionManagerPool = isUdRequest ? self.udSessionManagerPool : self.nonUdSessionManagerPool
        let sessionManager = sessionManagerPool.get()
        defer {
            sessionManagerPool.returnToPool(sessionManager)
        }

        let result = try await sessionManager.performRequest(request)

#if TESTABLE_BUILD
        if DebugFlags.logCurlOnSuccess {
            HTTPUtils.logCurl(for: request as URLRequest)
        }
#endif

        OutageDetection.shared.reportConnectionSuccess()

        return result
    }

    func makePromise(request: TSRequest) -> Promise<HTTPResponse> {
        return Promise.wrapAsync { return try await self.asyncRequest(request) }
    }

    func asyncRequest(_ request: TSRequest) async throws -> HTTPResponse {
        return try await makeRequest(request)
    }
}

// MARK: -

private class RESTSessionManager {

    private let urlSession: OWSURLSessionProtocol
    let createdDate = MonotonicDate()

    init() {
        urlSession = SSKEnvironment.shared.signalServiceRef.urlSessionForMainSignalService()
    }

    public func performRequest(_ request: TSRequest) async throws -> any HTTPResponse {
        // We should only use the RESTSessionManager for requests to the Signal main service.
        let urlSession = self.urlSession
        owsAssertDebug(urlSession.unfrontedBaseUrl == URL(string: TSConstants.mainServiceIdentifiedURL))

        do {
            return try await urlSession.performRequest(request)
        } catch let httpError as OWSHTTPError {
            // OWSUrlSession should only throw OWSHTTPError or OWSAssertionError.
            HTTPUtils.applyHTTPError(httpError)

            if httpError.httpStatusCode == 401, request.shouldCheckDeregisteredOn401 {
                try await makeIsDeregisteredRequest()
            }
            throw httpError
        } catch let error as CancellationError {
            throw error
        } catch {
            owsFailDebug("Unexpected error: \(error)")
            throw OWSHTTPError.invalidRequest
        }
    }

    private func makeIsDeregisteredRequest() async throws(CancellationError) {
        let isDeregisteredRequest = WhoAmIRequestFactory.amIDeregisteredRequest()

        let result: WhoAmIRequestFactory.Responses.AmIDeregistered?
        do {
            let response = try await self.performRequest(isDeregisteredRequest)
            result = WhoAmIRequestFactory.Responses.AmIDeregistered(rawValue: response.responseStatusCode)
        } catch let error as OWSHTTPError {
            result = WhoAmIRequestFactory.Responses.AmIDeregistered(rawValue: error.responseStatusCode)
        } catch let error as CancellationError {
            throw error
        } catch {
            result = nil
        }

        switch result {
        case .deregistered:
            Logger.warn("AmIDeregistered response says we are deregistered, marking as such.")
            await DependenciesBridge.shared.db.awaitableWrite { tx in
                DependenciesBridge.shared.registrationStateChangeManager.setIsDeregisteredOrDelinked(true, tx: tx)
            }
        case .notDeregistered:
            Logger.info("AmIDeregistered response says not deregistered; account probably disabled. Doing nothing.")
        case .none, .unexpectedError:
            Logger.error("Got unexpected AmIDeregistered response. Doing nothing.")
        }
    }
}

// MARK: -

// Session managers are stateful (e.g. the headers in the requestSerializer).
// Concurrent requests can interfere with each other. Therefore we use a pool
// do not re-use a session manager until its request succeeds or fails.
private class OWSSessionManagerPool {
    private let maxSessionManagerAge = 5 * 60 * NSEC_PER_SEC

    private let pool = AtomicValue<[RESTSessionManager]>([], lock: .init())

    // accessed from both networkManagerQueue and the main thread so needs a lock
    @Atomic private var lastDiscardDate: MonotonicDate?

    init() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(OWSSessionManagerPool.isCensorshipCircumventionActiveDidChange),
            name: Notification.Name.isCensorshipCircumventionActiveDidChange,
            object: nil)
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(OWSSessionManagerPool.isSignalProxyReadyDidChange),
            name: Notification.Name.isSignalProxyReadyDidChange,
            object: nil)
    }

    @objc
    private func isCensorshipCircumventionActiveDidChange() {
        AssertIsOnMainThread()
        lastDiscardDate = MonotonicDate()
    }

    @objc
    private func isSignalProxyReadyDidChange() {
        AssertIsOnMainThread()
        lastDiscardDate = MonotonicDate()
    }

    func get() -> RESTSessionManager {
        // Iterate over the pool, discarding expired session managers
        // until we find an unexpired session manager in the pool or
        // drain the pool and create a new session manager.
        while true {
            guard let sessionManager = pool.update(block: { $0.popLast() }) else {
                return RESTSessionManager()
            }
            if shouldDiscardSessionManager(sessionManager) {
                continue
            }
            return sessionManager
        }
    }

    func returnToPool(_ sessionManager: RESTSessionManager) {
        if shouldDiscardSessionManager(sessionManager) {
            return
        }
        let maxPoolSize = CurrentAppContext().isNSE ? 5 : 32
        pool.update {
            if $0.count < maxPoolSize {
                $0.append(sessionManager)
            }
        }
    }

    private func shouldDiscardSessionManager(_ sessionManager: RESTSessionManager) -> Bool {
        if let lastDiscardDate, sessionManager.createdDate < lastDiscardDate {
            return true
        }
        return (MonotonicDate() - sessionManager.createdDate) > maxSessionManagerAge
    }
}
